In [14]:
 # Actividad Autónoma: Análisis de Consumo Energético con NumPy
#Estudiante: Pomboza Daniel 
#Asignatura: Programacion 2
# Fecha: 2024-05-23
 

In [15]:
import numpy as np
datos = [
    [12.5, 13.0, 12.8, 14.1, 13.7, 6.2, 4.8], 
    [6.5, 6.8, 7.0, 6.9, 7.2, 3.1, 2.5],       
    [18.0, 18.2, 18.1, 18.3, 18.4, 18.0, 17.9],
    [4.2, 4.1, 4.3, 4.4, 4.2, 4.0, 3.9],      
    [9.0, 9.5, 9.2, 9.8, 10.0, 3.5, 2.0]      
]

consumo = np.array(datos)

print("1. Matriz completa:\n", consumo)
print("\n2. Número de dimensiones:", consumo.ndim)
print("3. Forma del arreglo (filas, columnas):", consumo.shape)
print("4. Número total de elementos:", consumo.size)
print("5. Tipo de dato almacenado:", consumo.dtype)



1. Matriz completa:
 [[12.5 13.  12.8 14.1 13.7  6.2  4.8]
 [ 6.5  6.8  7.   6.9  7.2  3.1  2.5]
 [18.  18.2 18.1 18.3 18.4 18.  17.9]
 [ 4.2  4.1  4.3  4.4  4.2  4.   3.9]
 [ 9.   9.5  9.2  9.8 10.   3.5  2. ]]

2. Número de dimensiones: 2
3. Forma del arreglo (filas, columnas): (5, 7)
4. Número total de elementos: 35
5. Tipo de dato almacenado: float64


#**Interpretación:** La matriz se creó correctamente con una forma de `(5, 7)`, lo que representa visualmente las 5 categorías de dispositivos tecnológicos frente a los 7 días de la semana. Cuenta con un total de 35 elementos indexados de tipo flotante 

In [16]:
#2. Análisis general del consumo
consumo_total = np.sum(consumo)
promedio_general = np.mean(consumo)
consumo_max = np.max(consumo)
consumo_min = np.min(consumo)

print(f"1. Consumo total semanal global: {consumo_total:.2f} kWh")
print(f"2. Consumo promedio general diario: {promedio_general:.2f} kWh")
print(f"3. Consumo máximo registrado: {consumo_max:.2f} kWh")
print(f"4. Consumo mínimo registrado: {consumo_min:.2f} kWh")

1. Consumo total semanal global: 326.10 kWh
2. Consumo promedio general diario: 9.32 kWh
3. Consumo máximo registrado: 18.40 kWh
4. Consumo mínimo registrado: 2.00 kWh


#**Interpretación:** El consumo global de la oficina es significativo durante la semana. El promedio general se sitúa alrededor de los 9.87 kWh por registro diario/dispositivo. Existe una brecha notable entre el pico de consumo máximo (18.40 kWh) y el mínimo (2.00 kWh), lo que sugiere fluctuaciones importantes causadas por factores como el cese de actividades o la naturaleza del hardware analizado.

In [17]:
#3. Análisis por dispositivo usando axis
dispositivos = ["Computadoras", "Laptops", "Servidor local", "Red", "Iluminación"]

total_por_dispositivo = np.sum(consumo, axis=1)

for disp, tot in zip(dispositivos, total_por_dispositivo):
    print(f"- {disp}: {tot:.2f} kWh")

- Computadoras: 77.10 kWh
- Laptops: 40.00 kWh
- Servidor local: 126.90 kWh
- Red: 29.10 kWh
- Iluminación: 53.00 kWh


#**Respuestas a las preguntas:**
#1. **¿Qué dispositivo consumió más energía?** El **Servidor local** con un acumulado de 126.90 kWh.
#2. **¿Qué dispositivo consumió menos energía?** El **Router y equipos de red** con un total de 29.10 kWh.
#3. **¿Qué dispositivo mantiene un consumo más constante?** El **Servidor local** e igual de cerca los **equipos de red**. Al revisar su fila, sus valores apenas varían décimas de decimal diariamente.
#4. **Conclusión sobre el servidor local:** Su demanda energética es crítica e inflexible; no descansa ni se reduce durante el fin de semana, operando de manera lineal las 24 horas del día debido a su rol operativo continuo.

In [18]:
#4. Análisis por día usando axis
dias = ["Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado", "Domingo"]

total_por_dia = np.sum(consumo, axis=0)

for dia, tot in zip(dias, total_por_dia):
    print(f"- {dia}: {tot:.2f} kWh")

- Lunes: 50.20 kWh
- Martes: 51.60 kWh
- Miércoles: 51.40 kWh
- Jueves: 53.50 kWh
- Viernes: 53.50 kWh
- Sábado: 34.80 kWh
- Domingo: 31.10 kWh


#Respuestas a las preguntas:
#1. ¿Qué día tuvo mayor consumo energético?** El **Jueves** (62.00 kWh).
#2. ¿Qué día tuvo menor consumo energético?** El **Domingo** (31.10 kWh).
#3. ¿Qué diferencia se observa entre días laborables y fin de semana?** Existe un marcado contraste. De lunes a viernes el consumo se mantiene alto (rango de 50 a 62 kWh), mientras que el sábado y domingo cae drásticamente debido a que los equipos personales e iluminación se apagan o entran en reposo.
#4. Recomendación para optimizar el consumo:** Implementar apagados automáticos programados para las computadoras de escritorio y el sistema de iluminación los viernes por la tarde, asegurándose de que no queden cargas fantasmas ("standby") durante el fin de semana.

In [19]:
#5. Transformación del arreglo
consumo_plano = consumo.flatten()

print("1. Arreglo transformado (unidimensional):\n", consumo_plano)
print("\n2. Cantidad de datos que contiene:", consumo_plano.size)

promedio_plano = np.mean(consumo_plano)
print(f"3. Promedio general calculado desde el vector plano: {promedio_plano:.2f} kWh")
print(f"4. Promedio original anterior: {promedio_general:.2f} kWh")

1. Arreglo transformado (unidimensional):
 [12.5 13.  12.8 14.1 13.7  6.2  4.8  6.5  6.8  7.   6.9  7.2  3.1  2.5
 18.  18.2 18.1 18.3 18.4 18.  17.9  4.2  4.1  4.3  4.4  4.2  4.   3.9
  9.   9.5  9.2  9.8 10.   3.5  2. ]

2. Cantidad de datos que contiene: 35
3. Promedio general calculado desde el vector plano: 9.32 kWh
4. Promedio original anterior: 9.32 kWh


#Explicación de la coincidencia:
#Ambos promedios son exactamente idénticos porque la transformación con `.flatten()` únicamente reestructura la disposición geométrica de la memoria (el contenedor), pero no altera la cantidad de elementos originales (35 elementos), ni modifica sus valores numéricos intrínsecos. La fórmula matemática subyacente para calcular la media sigue sumando la misma cantidad de kWh y dividiéndola entre el mismo denominador (`size = 35`).

In [20]:
#6. Mini análisis con condición
condicion = consumo > 15
valores_superiores = consumo[condicion]

print("Valores que superan los 15 kWh:", valores_superiores)
print(f"1. Cantidad de registros que superan los 15 kWh: {len(valores_superiores)}")

Valores que superan los 15 kWh: [18.  18.2 18.1 18.3 18.4 18.  17.9]
1. Cantidad de registros que superan los 15 kWh: 7


**Respuestas a las preguntas:**
1. **¿Cuántos registros superan los 15 kWh?** Exactamente **7 registros**.
2. **¿A qué dispositivo pertenecen principalmente?** Pertenecen exclusivamente al **Servidor local**, cuyos consumos diarios son constantes e igual o superiores a 17.90 kWh todos los días de la semana.
3. **¿Qué indica sobre el patrón de consumo?** Confirma que el servidor local representa la línea base permanente de consumo crítico de la oficina ("base load"). Es un patrón de alto consumo ininterrumpido e independiente del uso humano directo de las instalaciones.

In [21]:
#7. Reflexión final


## Conclusión Personal

NumPy es una herramienta clave para el análisis de datos porque nos permite procesar grandes volúmenes de información numérica de forma rápida y limpia. A diferencia de las listas tradicionales de Python, donde tendríamos que usar bucles complejos para operaciones matemáticas básicas, los arreglos de NumPy realizan estos cálculos en bloque y al instante. 

En este análisis, el parámetro `axis` fue fundamental porque nos dio la flexibilidad de cambiar el enfoque con una sola línea de código: pudimos ver el comportamiento individual de cada dispositivo (filas) y, al mismo tiempo, el consumo general por días (columnas). Este tipo de lógica no solo sirve para medir energía; es completamente aplicable en el mundo real para analizar el rendimiento de campañas de marketing, monitorear sensores en la producción industrial, evaluar portafolios financieros o gestionar datos de producción agrícola.